### ISSUE TITLE
Server startup fails with emacs-lsp

### ISSUE THREAD

**Issue by thegeorgeous** [castwide/solargraph#460]
When using solargraph as the lsp server for emacs lsp, `solargraph stdio` fails with the error below:
```
Solargraph is listening on stdio PID=97310
#<Thread:0x00000001400762a0@/Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/backport-1.2.0/lib/backport/client.rb:104 run> terminated with exception (report_on_exception is true):
/Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/2.6.0/ripper/sexp.rb:80:in `===': stack level too deep (SystemStackError)
	from /Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/2.6.0/ripper/sexp.rb:80:in `block (2 levels) in on_heredoc_dedent'
	from /Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/2.6.0/ripper/sexp.rb:75:in `map!'
	from /Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/2.6.0/ripper/sexp.rb:75:in `block in on_heredoc_dedent'
	from /Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/2.6.0/ripper/sexp.rb:81:in `block (2 levels) in on_heredoc_dedent'
	from /Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/2.6.0/ripper/sexp.rb:75:in `map!'
	from /Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/2.6.0/ripper/sexp.rb:75:in `block in on_heredoc_dedent'
	from /Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/2.6.0/ripper/sexp.rb:81:in `block (2 levels) in on_heredoc_dedent'
	from /Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/2.6.0/ripper/sexp.rb:75:in `map!'
	 ... 370 levels...
	from /Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/2.6.0/observer.rb:196:in `each'
	from /Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/2.6.0/observer.rb:196:in `notify_observers'
	from /Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/backport-1.2.0/lib/backport/client.rb:124:in `read_input'
	from /Users/thegeorgeous/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/backport-1.2.0/lib/backport/client.rb:105:in `block in run_input_thread'
```

**Comment by castwide** [comment]
Does it happen on a specific workspace? Based on the stacktrace, it looks like Ripper is failing to parse a heredoc string in one of your Ruby files. (Solargraph uses Ripper to extract comments.)

**Comment by thegeorgeous** [comment]
@castwide

`solargraph scan -v` completed without any issues
```
Scanned /Users/thegeorgeous/Workspace/application (105019 pins) in 331.13377199997194 seconds.
```

Is there some other way I can find this rogue heredoc string?

**Comment by thegeorgeous** [comment]
@castwide 

The issue was because of a squiggly heredoc string used to write a SQL query. I was able to run `Ripper.sexp` on the string but it seems to be failing in the server. 

You can close this ticket if it is not a solargraph-related issue.

**Comment by castwide** [comment]
If you can give me an example of a heredoc string that causes the error, I can try to make it fail more gracefully or look for some other workaround.

**Comment by thegeorgeous** [comment]
@castwide, I'm afraid I can't share the SQL query as it is part of proprietary code. I can tell that it does string interpolation in two places. I tried removing the interpolation in case that was the issue, but that didn't help. 

The only way to make it work was to change it from `<<~SQL` to `<<-SQL`

**Comment by castwide** [comment]
Understood, thanks. I'll keep this issue open for at least a little longer while I try to reproduce the error.

**Comment by il-tmfv** [comment]
I'm having the same error if I run solargraph manually + `external` transport:
```
#<Thread:0x00000001571bae60 /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/backport-1.2.0/lib/backport/client.rb:104 run> terminated with exception (report_on_exception is true):
Traceback (most recent call last):
	3194: from /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/backport-1.2.0/lib/backport/client.rb:105:in `block in run_input_thread'
	3193: from /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/backport-1.2.0/lib/backport/client.rb:124:in `read_input'
	3192: from /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/2.7.0/observer.rb:196:in `notify_observers'
	3191: from /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/2.7.0/observer.rb:196:in `each'
	3190: from /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/2.7.0/observer.rb:197:in `block in notify_observers'
	3189: from /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/backport-1.2.0/lib/backport/server/tcpip.rb:82:in `update'
	3188: from /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/backport-1.2.0/lib/backport/client.rb:66:in `tick'
	3187: from /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/solargraph-0.43.0/lib/solargraph/language_server/transport/adapter.rb:27:in `receiving'
	 ... 3182 levels...
	   4: from /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/2.7.0/ripper/sexp.rb:75:in `block in on_heredoc_dedent'
	   3: from /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/2.7.0/ripper/sexp.rb:75:in `map!'
	   2: from /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/2.7.0/ripper/sexp.rb:81:in `block (2 levels) in on_heredoc_dedent'
	   1: from /Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/2.7.0/ripper/sexp.rb:75:in `block in on_heredoc_dedent'
/Users/timofeev/.rbenv/versions/2.7.2/lib/ruby/2.7.0/ripper/sexp.rb:75:in `map!': stack level too deep (SystemStackError)
```

In the case of VS code integration it does nothing: no errors, all commands fail (like getting gem version).

I've tried to find what causes it and looks like I've found:
```ruby
class MigrateXxx < ActiveRecord::Migration[6.1]
  def up
    return if Rails.env.development? || Rails.env.test?
    execute <<~SQL
      delete from place_features where
        feature_id in (1, 2) -- some comment
        and
        place_id in (
          1,
          2,
          -- !!! HERE will be 3, 4, ... till ~3800
          3800
        );
    SQL
  end

  def down
    # noop
  end
end

```

It is a very long `SQL` string with more than 3800 lines 😅 

Changing `<<~SQL` to `<<-SQL` makes everything work. As well as removing a lot of lines.

Hope it helps. Luckily I can ignore migration files in config.

**Comment by castwide** [comment]
@il-tmfv Thanks for the example. I was able to reproduce the error with that list expanded to 3800 lines. I'll look into a solution.

**Comment by castwide** [comment]
Fixed in v0.43.1.

**Comment by thegeorgeous** [comment]
@castwide the issue is resolved for me too. I am closing the issue. 🥳
